lib/http2.c

```diff
@@ -2139,23 +2184,61 @@ static ssize_t cf_h2_send(struct Curl_cfilter *cf, struct Curl_easy *data,
 
 out:
   if(stream) {
+    if(nwritten > 0 && !Curl_bufq_is_empty(&ctx->outbufq) &&
+        !stream->upload_left) {
+      CURLcode rv2;
+      int i;
+      DEBUGF(LOG_CF(data, cf, "[h2sid=%d] cf_send(len=%zu) of last upload "
+                    "chunk was not complete (%zu buffered), "
+                    "using timed write to finish",
+                    stream->id, len, Curl_bufq_len(&ctx->outbufq)));
+      for(i = 0; i < 1000; ++i) {
+        rv2 = nw_out_flush(cf, data);
+        if(!rv2) {
+          DEBUGF(LOG_CF(data, cf, "[h2sid=%d] cf_send() out flushed",
+                        stream->id));
+          break;
+        }
+        else if(rv2 == CURLE_AGAIN) {
+          DEBUGF(LOG_CF(data, cf, "[h2sid=%d] cf_send() pausing "
+                        "10ms before next flush attempt",
+                        stream->id));
+          Curl_wait_ms(10);
+        }
+        else {
+          *err = rv2;
+          nwritten = -1;
+          goto last_out;
+        }
+      }
+      if(rv2 == CURLE_AGAIN) {
+        failf(data, "[HTTP/2] [h2sid=%d] cf_send() failed to flush "
+                      "nw out buffer of %zu bytes",
+                      stream->id, Curl_bufq_len(&ctx->outbufq));
+      }
+    }
     DEBUGF(LOG_CF(data, cf, "[h2sid=%d] cf_send(len=%zu) -> %zd, %d, "
-                  "buffered=%zu, upload_left=%zu, stream-window=%d, "
-                  "connection-window=%d",
+                  "buffered=%zu, upload_left=%" CURL_FORMAT_CURL_OFF_T
+                  ", stream-window=%d, "
+                  "connection-window=%d, stream_send_buffer(%zu), "
+                  "nw_send_buffer(%zu)",
                   stream->id, len, nwritten, *err,
                   Curl_bufq_len(&stream->sendbuf),
                   (ssize_t)stream->upload_left,
```

Comment:
the typecast needs to be removed if you are using CURL_FORMAT_CURL_OFF_T
```suggestion
                  stream->upload_left,
```